File: src/backtracing/_494_find_target_sum_ways.rs
struct Solution;

impl Solution {
    pub fn find_target_sum_ways(nums: Vec<i32>, target: i32) -> i32 {
        fn dfs(nums: &[i32], target: i32, index: usize, sum: i32, ans: &mut i32) {
            if index == nums.len() {
                if sum == target {
                    *ans += 1;
                }
                return;
            }
            dfs(nums, target, index + 1, sum + nums[index], ans);
            dfs(nums, target, index + 1, sum - nums[index], ans);
        }
        let mut ans = 0_i32;
        dfs(&nums, target, 0, 0, &mut ans);
        ans
    }
}

File: leetcode-go/494_findTargetSumWays.go
package leetcode_go

func findTargetSumWays(nums []int, target int) int {
	sum := 0
	for _, num := range nums {
		sum += num
	}

	if target > sum || (sum+target)%2 == 1 {
		return 0
	}

	bag := (sum + target) / 2

	dp := make([]int, bag+1)
	dp[0] = 1

	for i := 0; i < len(nums); i++ {
		for j := bag; j >= nums[i]; j++ {
			dp[j] += dp[j-nums[i]]
		}
	}

	return dp[bag]
}
